# Email Spam Filter

Aim: Build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. 
To classify messages as spam or non-spam the computer:

* Learns how humans classify messages.
* Uses that human knowledge to estimate probabilities for new messages — probabilities for spam and non-spam.
* Classifies a new message based on these probability values — if the probability for spam is greater, then it classifies the message as spam. Otherwise, it classifies it as non-spam (if the two probability values are equal, then we may need a human to classify the message).

The program should classifies new messages with an accuracy greater than 80%. Therefore, expecting that more than 80% of the new messages will be classified correctly as spam or ham (non-spam).

Keywords: Naive Bayes Classifier, Machine Learning, Feature Engineering, Data Cleaning

Dataset: To train the algorithm, the dataset consists of 5,572 SMS messages that are already classified by humans. The dataset was put together by Tiago A. Almeida and José María Gómez Hidalgo, and it can be downloaded from the [The UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/sms+spam+collection).


## Exploratory  Data Analysis

In [2]:
import pandas as pd

# The data points are tab separated, so need to use the sep='\t' parameter for our read_csv() function.
# The dataset doesn't have a header row, which means using the header=None parameter, otherwise the first row will be wrongly used as the header row.
# Use the names=['Label', 'SMS'] parameter to name the columns as Label and SMS
sms_spam = pd.read_csv('SMSSpamCollection', sep='\t', header=None, names=['Label', 'SMS'])

print(sms_spam.shape)
sms_spam.head()

(5572, 2)


Label                                                SMS
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

As shown below, a estimate 87% of the messages are ham, and the remaining 13% are spam. This sample looks representative, since in practice most messages that people receive are ham.

In [3]:
sms_spam['Label'].value_counts(normalize=True)

Label
ham     0.865937
spam    0.134063
Name: proportion, dtype: float64

## Training and Test Set

Split our dataset into a training and a test set, where the training set accounts for 80% of the data, and the test set for the remaining 20%.

In [4]:
# Randomize the dataset
data_randomized = sms_spam.sample(frac=1, random_state=1)

# Calculate index for split
training_test_index = round(len(data_randomized) * 0.8)

# Training/Test split
training_set = data_randomized[:training_test_index].reset_index(drop=True)
test_set = data_randomized[training_test_index:].reset_index(drop=True)

print(training_set.shape)
print(test_set.shape)

(4458, 2)
(1114, 2)


Analyze the percentage of spam and ham messages in the training and test sets. The percentages should be close to what is in the full dataset, where about 87% of the messages are ham, and the remaining 13% are spam.

In [5]:
training_set['Label'].value_counts(normalize=True)

Label
ham     0.86541
spam    0.13459
Name: proportion, dtype: float64

In [6]:
test_set['Label'].value_counts(normalize=True)

Label
ham     0.868043
spam    0.131957
Name: proportion, dtype: float64

## Data Cleaning

To calculate all the probabilities required by the algorithm, performed data cleaning to bring the data in a format that will allow all the needed information to be extracted easily.

The data should be transformed into this format:

![img](https://dq-content.s3.amazonaws.com/433/cpgp_dataset_3.png)


### Letter Case and Punctuation

Removing all the punctuation and bringing every letter to lower case.

In [7]:
# Before cleaning
training_set.head()

Label                                                SMS
0   ham                       Yep, by the pretty sculpture
1   ham      Yes, princess. Are you going to make me moan?
2   ham                         Welp apparently he retired
3   ham                                            Havent.
4   ham  I forgot 2 ask ü all smth.. There's a card on ...

In [8]:
# After cleaning
training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')
training_set['SMS'] = training_set['SMS'].str.lower()
training_set.head()

<>:2: SyntaxWarning: invalid escape sequence '\W'
<>:2: SyntaxWarning: invalid escape sequence '\W'
C:\Users\eilee\AppData\Local\Temp\ipykernel_14924\3773386248.py:2: SyntaxWarning: invalid escape sequence '\W'
  training_set['SMS'] = training_set['SMS'].str.replace('\W', ' ')


Label                                                SMS
0   ham                       yep, by the pretty sculpture
1   ham      yes, princess. are you going to make me moan?
2   ham                         welp apparently he retired
3   ham                                            havent.
4   ham  i forgot 2 ask ü all smth.. there's a card on ...

### Creating the Vocabulary

Creating a list with all the unique words in the training set, where each word is represented as a string.

In [9]:
training_set['SMS'] = training_set['SMS'].str.split()

vocabulary = []
for sms in training_set['SMS']:
    for word in sms:
        vocabulary.append(word)
        
vocabulary = list(set(vocabulary))
len(vocabulary)

11860

There are 7,783 unique words in all the messages of the training set.

### Final Training Set

The vocabulary list is used to make data transformation of the required data.

In [10]:
# Initialise a dictionary where each key is a unique word from the vocabulary, and each value is a list of the length of training set, where each element in the list is a 0.
word_counts_per_sms = {unique_word: [0] * len(training_set['SMS']) for unique_word in vocabulary}

# Loop over training_set['SMS'] to get both the index and the SMS message 
for index, sms in enumerate(training_set['SMS']):
    # Loop over each text
    for word in sms:
        # Counts the frequency
        word_counts_per_sms[word][index] += 1

In [11]:
word_counts = pd.DataFrame(word_counts_per_sms)
word_counts.head()

sport  +std  you‘ll  lipo.  sleep..  reache  nurses  duo  rest?  now.  ...  \
0      0     0       0      0        0       0       0    0      0     0  ...   
1      0     0       0      0        0       0       0    0      0     0  ...   
2      0     0       0      0        0       0       0    0      0     0  ...   
3      0     0       0      0        0       0       0    0      0     0  ...   
4      0     0       0      0        0       0       0    0      0     0  ...   

   today...  granted  wrking  boyfriend  ok.  plum  stretch  l-oveable.  \
0         0        0       0          0    0     0        0           0   
1         0        0       0          0    0     0        0           0   
2         0        0       0          0    0     0        0           0   
3         0        0       0          0    0     0        0           0   
4         0        0       0          0    0     0        0           0   

   official  mrng:-)  
0         0        0  
1         0        0  
2         0        0  
3         0        0  
4         0        0  

[5 rows x 11860 columns]

Missing full messages and labels, therefore, added two columns to word_counts

In [12]:
training_set_clean = pd.concat([training_set, word_counts], axis=1)
training_set_clean.head()

Label                                                SMS  sport  +std  \
0   ham                 [yep,, by, the, pretty, sculpture]      0     0   
1   ham  [yes,, princess., are, you, going, to, make, m...      0     0   
2   ham                    [welp, apparently, he, retired]      0     0   
3   ham                                          [havent.]      0     0   
4   ham  [i, forgot, 2, ask, ü, all, smth.., there's, a...      0     0   

   you‘ll  lipo.  sleep..  reache  nurses  duo  ...  today...  granted  \
0       0      0        0       0       0    0  ...         0        0   
1       0      0        0       0       0    0  ...         0        0   
2       0      0        0       0       0    0  ...         0        0   
3       0      0        0       0       0    0  ...         0        0   
4       0      0        0       0       0    0  ...         0        0   

   wrking  boyfriend  ok.  plum  stretch  l-oveable.  official  mrng:-)  
0       0          0    0     0        0           0         0        0  
1       0          0    0     0        0           0         0        0  
2       0          0    0     0        0           0         0        0  
3       0          0    0     0        0           0         0        0  
4       0          0    0     0        0           0         0        0  

[5 rows x 11862 columns]

## Calculating Constants First

The Naive Bayes algorithm will need to answer these two probability questions to be able to classify new messages:

\begin{equation}
P(Spam | w_1,w_2, ..., w_n) \propto P(Spam) \cdot \prod_{i=1}^{n}P(w_i|Spam)
\end{equation}

\begin{equation}
P(Ham | w_1,w_2, ..., w_n) \propto P(Ham) \cdot \prod_{i=1}^{n}P(w_i|Ham)
\end{equation}


Also, to calculate P(w<sub>i</sub>|Spam) and P(w<sub>i</sub>|Ham) inside the formulas above, these equations were used:

\begin{equation}
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
\end{equation}

\begin{equation}
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
\end{equation}


Some of the terms in the four equations above will have the same value for every new message. Able to calculate the value of these terms once and avoid doing the computations again when a new messages comes in. Below, the cleaned training set is used to calculate:

- P(Spam) and P(Ham)
- N<sub>Spam</sub>, N<sub>Ham</sub>, N<sub>Vocabulary</sub>

Also use Laplace smoothing and set $\alpha = 1$.

In [13]:
# Isolating spam and ham messages first
spam_messages = training_set_clean[training_set_clean['Label'] == 'spam']
ham_messages = training_set_clean[training_set_clean['Label'] == 'ham']

# P(Spam) and P(Ham)
p_spam = len(spam_messages) / len(training_set_clean)
p_ham = len(ham_messages) / len(training_set_clean)

# N_Spam
n_words_per_spam_message = spam_messages['SMS'].apply(len)
n_spam = n_words_per_spam_message.sum()

# N_Ham
n_words_per_ham_message = ham_messages['SMS'].apply(len)
n_ham = n_words_per_ham_message.sum()

# N_Vocabulary
n_vocabulary = len(vocabulary)

# Laplace smoothing
alpha = 1

## Calculating Parameters

With the constant terms calculated above, the next step is to calculate the parameters $P(w_i|Spam)$ and $P(w_i|Ham)$. Each parameter will thus be a conditional probability value associated with each word in the vocabulary.

The parameters are calculated using the formulas:

\begin{equation}
P(w_i|Spam) = \frac{N_{w_i|Spam} + \alpha}{N_{Spam} + \alpha \cdot N_{Vocabulary}}
\end{equation}

\begin{equation}
P(w_i|Ham) = \frac{N_{w_i|Ham} + \alpha}{N_{Ham} + \alpha \cdot N_{Vocabulary}}
\end{equation}

In [14]:
# Initiate parameters
parameters_spam = {unique_word:0 for unique_word in vocabulary}
parameters_ham = {unique_word:0 for unique_word in vocabulary}

# Calculate parameters
for word in vocabulary:
    n_word_given_spam = spam_messages[word].sum()   # spam_messages already defined in a cell above
    p_word_given_spam = (n_word_given_spam + alpha) / (n_spam + alpha*n_vocabulary)
    parameters_spam[word] = p_word_given_spam
    
    n_word_given_ham = ham_messages[word].sum()   # ham_messages already defined in a cell above
    p_word_given_ham = (n_word_given_ham + alpha) / (n_ham + alpha*n_vocabulary)
    parameters_ham[word] = p_word_given_ham

## Classifying A New Message

With the parameteres calculated, the spam filter can be understood as a function that:

- Takes in as input a new message (w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>).
- Calculates P(Spam|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>) and P(Ham|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>).
- Compares the values of P(Spam|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>) and P(Ham|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>), and:
    - If P(Ham|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>) > P(Spam|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>), then the message is classified as ham.
    - If P(Ham|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>) < P(Spam|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>), then the message is classified as spam.
    -  If P(Ham|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>) = P(Spam|w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>), then the algorithm may request human help.

In [15]:
import re

def classify(message):
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
            
    print('P(Spam|message):', p_spam_given_message)
    print('P(Ham|message):', p_ham_given_message)
    
    if p_ham_given_message > p_spam_given_message:
        print('Label: Ham')
    elif p_ham_given_message < p_spam_given_message:
        print('Label: Spam')
    else:
        print('Equal proabilities, have a human classify this!')

<>:8: SyntaxWarning: invalid escape sequence '\W'
<>:8: SyntaxWarning: invalid escape sequence '\W'
C:\Users\eilee\AppData\Local\Temp\ipykernel_14924\3418688428.py:8: SyntaxWarning: invalid escape sequence '\W'
  message = re.sub('\W', ' ', message)


In [16]:
classify('WINNER!! This is the secret code to unlock the money: C3421.')

P(Spam|message): 1.1680023632078457e-26
P(Ham|message): 6.088544142463393e-28
Label: Spam


In [17]:
classify("Sounds good, Tom, then see u there")

P(Spam|message): 2.234299283967944e-26
P(Ham|message): 8.376346103813855e-22
Label: Ham


## Measuring the Spam Filter's Accuracy

Compared the train dataset with the test dataset.

In [18]:
# Returns the classification labels
def classify_test_set(message):    
    '''
    message: a string
    '''
    
    message = re.sub('\W', ' ', message)
    message = message.lower().split()
    
    p_spam_given_message = p_spam
    p_ham_given_message = p_ham

    for word in message:
        if word in parameters_spam:
            p_spam_given_message *= parameters_spam[word]
            
        if word in parameters_ham:
            p_ham_given_message *= parameters_ham[word]
    
    if p_ham_given_message > p_spam_given_message:
        return 'ham'
    elif p_spam_given_message > p_ham_given_message:
        return 'spam'
    else:
        return 'needs human classification'

<>:7: SyntaxWarning: invalid escape sequence '\W'
<>:7: SyntaxWarning: invalid escape sequence '\W'
C:\Users\eilee\AppData\Local\Temp\ipykernel_14924\4012572072.py:7: SyntaxWarning: invalid escape sequence '\W'
  message = re.sub('\W', ' ', message)


In [19]:
# With the function above, a new column is created in the test set.
test_set['predicted'] = test_set['SMS'].apply(classify_test_set)
test_set.head()

Label                                                SMS predicted
0   ham          Later i guess. I needa do mcat study too.       ham
1   ham             But i haf enuff space got like 4 mb...       ham
2  spam  Had your mobile 10 mths? Update to latest Oran...      spam
3   ham  All sounds good. Fingers . Makes it difficult ...       ham
4   ham  All done, all handed in. Don't know if mega sh...       ham

Measured the accuracy of the spam filter below.

In [20]:
correct = 0
total = test_set.shape[0]
    
for row in test_set.iterrows():
    row = row[1]
    if row['Label'] == row['predicted']:
        correct += 1
        
print('Correct:', correct)
print('Incorrect:', total - correct)
print('Accuracy:', correct/total)

Correct: 1090
Incorrect: 24
Accuracy: 0.9784560143626571


The accuracy is at least 97%. The spam filter scanned at 1,114 messages that it has not seen in training, and classified 1,100 correctly.

## Next Steps

In this project, managed to build a spam filter for SMS messages using the multinomial Naive Bayes algorithm. The filter had an accuracy of 98.74% on the test set, which met the initial goal of  an accuracy of over 80%.

Next steps include:

- Analyze the 14 messages that were classified incorrectly and try to figure out why the algorithm classified them incorrectly
- Make the filtering process more complex by making the algorithm sensitive to letter case